### PR TITLE
osr: direct frame delivery via `discord_overlay2.node`.

### PR DESCRIPTION
**Pull request by hinaria**

### CODE REVIEW

File: discord/overlay_win.cc [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+                               uint32_t height,
+                               void* data,
+                               size_t length) {
+  if (process_id_ == 0) {
+    return false;
+  } else {
+    auto frame = OverlayFrameData{process_id_, width, height, data, length};
+
+    auto send_frame =
+        FindOverlayFunction<bool (*)(uint32_t version, void* data)>(
```
Comment (vassjozsef) [inline discord/overlay_win.cc:57]: Do we need to call this for every single frame?

File: atom/browser/api/atom_api_web_contents.cc [hunk 2]
```diff
@@ -1854,7 +1854,10 @@ bool WebContents::IsOffScreen() const {
 
 #if BUILDFLAG(ENABLE_OSR)
 void WebContents::OnPaint(const gfx::Rect& dirty_rect, const SkBitmap& bitmap) {
-  Emit("paint", dirty_rect, gfx::Image::CreateFrom1xBitmap(bitmap));
+  if (!overlay_.SendFrame(bitmap.width(), bitmap.height(), bitmap.getPixels(),
+                          bitmap.rowBytes() * bitmap.height())) {
+    Emit("paint", dirty_rect, gfx::Image::CreateFrom1xBitmap(bitmap));
```
Comment (vassjozsef) [inline atom/browser/api/atom_api_web_contents.cc:1859]: Shall we still emit paint event even if we directly the frame to native module?

File: discord/overlay_win.cc [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+            send_function_loaded_, send_function_, "SendOverlayFrame");
+
+    return send_frame != nullptr &&
+           send_frame(OverlayFrameData::Version, &frame);
```
Comment (vassjozsef) [inline discord/overlay_win.cc:61]: Why do we need version twice?

File: discord/overlay_win.cc (hunk 1 above)
Comment (hinaria) [reply discord/overlay_win.cc]: `FindOverlayFunction` effectively becomes `if (dll_loaded) { return function_handle; }` once `discord_overlay2.node` is loaded for the first time so it's pretty cheap.

File: atom/browser/api/atom_api_web_contents.cc (hunk 2 above)
Comment (hinaria) [reply atom/browser/api/atom_api_web_contents.cc]: avoiding the js sync point is where like 80% of our performance win comes from, so i'd rather not unless there is a strong motivation to do so.

File: discord/overlay.h [hunk 4]
```diff
@@ -0,0 +1,17 @@
+#include <cstdint>
+
+namespace discord {
+
+class DiscordOverlay {
```
Comment (adill) [inline discord/overlay.h]: nit: `Overlay` is sufficient

File: discord/overlay.h [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+  bool SendFrame(uint32_t width, uint32_t height, void* data, size_t length);
+
+ private:
+  uint32_t process_id_ = 0;
+
+  bool send_function_loaded_ = false;
+  const void* send_function_ = nullptr;
+};
+
+}  // namespace discord
```
Comment (adill) [inline discord/overlay.h]: nit: missing trailing newline

File: discord/overlay_win.cc (hunk 3 above)
Comment (hinaria) [reply discord/overlay_win.cc]: `OverlayFrameData::Version` is an associated constant, not a struct member.

**Review by adill — APPROVED** [review]
lgtm 👍

File: discord/overlay.h (hunk 5 above)
Comment (hinaria) [reply discord/overlay.h]: hm, `clang-format` doesn't automatically add trailing new lines?

**Pull request merged**